### PR TITLE
ZEN-28644: Defer generating object specs until the end of the transac…

### PR DESCRIPTION
**Pull request by fraibacas**
…tion

### CODE REVIEW

File: Products/Zuul/catalog/model_catalog.py [hunk 1]
```diff
@@ -263,49 +261,66 @@ def add_model_update(self, object_update):
         if idxs: # make sure mandatory idxs are always sent
             idxs.update( MANDATORY_FIELDS ) # Mandatory fields
 
-        model_update = IndexUpdate(object_update.obj, op=op , idxs=idxs, uid=uid)
-        self.pending_updates[object_update.uid] = model_update
-        del object_update # Make sure we dont keep references to the object
```
Comment (johnlhamilton) [inline Products/Zuul/catalog/model_catalog.py]: should we still do this?

**Comment by fraibacas** [comment]
this is a good way to test these changes:
https://gist.github.com/fraibacas/535e81046dd0f254f7f389d2c638f683

File: Products/Zuul/catalog/model_catalog.py (hunk 1 above)
Comment (fraibacas) [reply Products/Zuul/catalog/model_catalog.py]: I think it was redundant and the python garbage collector will do its job

File: Products/Zuul/catalog/model_catalog.py [hunk 2]
```diff
@@ -337,18 +352,18 @@ def get_indexes(self):
...
+        """ index all pending updates during a mid transaction commit """
         updates = tx_state.get_pending_updates()
         # we are going to index all pending updates setting the MODEL_INDEX_UID_FIELD
         # field as the OBJECT_UID_FIELD appending the tid and setting tx_state field
         # to the current tid
         tweaked_updates = []
         indexed_uids = set()
         deleted_uids = set()
-        for update in updates:
+        for update in updates.values():
```
Comment (iancmcc) [inline Products/Zuul/catalog/model_catalog.py]: `itervalues()`

File: Products/Zuul/catalog/model_catalog.py [hunk 3]
```diff
@@ -263,49 +261,66 @@ def add_model_update(self, object_update):
         if idxs: # make sure mandatory idxs are always sent
             idxs.update( MANDATORY_FIELDS ) # Mandatory fields
 
-        model_update = IndexUpdate(object_update.obj, op=op , idxs=idxs, uid=uid)
-        self.pending_updates[object_update.uid] = model_update
-        del object_update # Make sure we dont keep references to the object
+        # store the ObjectUpdate. Defer creating the IndexUpdate
+        # until the end of the transaction or before a dirty search
+        self.pending_updates[object_update.uid] = ObjectUpdate(object_update.obj, op=op , idxs=idxs)
```
Comment (iancmcc) [inline Products/Zuul/catalog/model_catalog.py]: That `del` is technically still valuable, although extra credit. But it's possible for that to leak. It should have been in the caller of this method, though.

**Pull request merged**